### PR TITLE
Implement fill tool

### PR DESCRIPTION
**Pull request by bttrs**
This PR aims to solve [#24](https://github.com/scribble-rs/scribble.rs/issues/24).

It completely changes how things are drawn onto the canvas, to prevent ugly anti-aliasing edges. It seems to be fine perfomance wise.

### CODE REVIEW

**Comment by Bios-Marcel** [comment]
Awesome, I'll test that thoroughly later

**Comment by Bios-Marcel** [comment]
Next time pls don't put the PR on your master branch, makes stuff easier for me :D

File: lobby.html [hunk 1]
```diff
@@ -397,8 +423,14 @@
             drawLine(context, parsed.Data.FromX * scaleDownFactor, parsed.Data.FromY * scaleDownFactor, parsed.Data.ToX * scaleDownFactor, parsed.Data.ToY * scaleDownFactor, parsed.Data.Color, parsed.Data.LineWidth * scaleDownFactor);
         } else if (parsed.Type === "pixels") {
             parsed.Data.forEach(function(pixel) {
-                drawLine(context, pixel.FromX * scaleDownFactor, pixel.FromY * scaleDownFactor, pixel.ToX * scaleDownFactor, pixel.ToY * scaleDownFactor, pixel.Color, pixel.LineWidth * scaleDownFactor);
+                if (pixel.Type === "fill") {
```
Comment (Bios-Marcel) [inline lobby.html]: Do we really need "pixels"."fill" and "fill"?

File: lobby.go [hunk 2]
```diff
@@ -222,6 +223,24 @@ func wsListen(lobby *Lobby, player *Player, socket *websocket.Conn) {
 						}
 					}
 				}
+			} else if received.Type == "fill" {
+				//TODO: append pixel data
```
Comment (Bios-Marcel) [inline lobby.go]: This doesn't seem to be necessary.

File: lobby.html [hunk 3]
```diff
@@ -310,38 +320,54 @@
...
+        subtree: false
+    });
 
     socket.onopen = () => {
         console.log("Successfully Connected");
     };
 
     socket.onmessage = event => {
         var parsed = JSON.parse(event.data);
+        console.log("event: " + parsed.Type);
```
Comment (Bios-Marcel) [inline lobby.html]: I'd prefer not unnecessarily having logger code on the client. If you need that for debugging purposes, you can simply add that real quick.

File: lobby.html [hunk 4]
```diff
@@ -533,7 +559,154 @@
...
+            }
+            if (e2 < dx) {
+                err += dx;
+                y0 += sy;
+            }
+        }
+    }
+
+    function generateCircleMap(radius) {
+        //console.log("generateCircleData: radius:" + radius);
```
Comment (Bios-Marcel) [inline lobby.html]: Remove

File: lobby.html [hunk 5]
```diff
@@ -533,7 +559,154 @@
...
+
+    function drawBresenhamLine(imageData, x0, y0, x1, y1, color) {
+        var dx = Math.abs(x1 - x0);
+        var dy = Math.abs(y1 - y0);
+        var sx = (x0 < x1) ? 1 : -1;
+        var sy = (y0 < y1) ? 1 : -1;
+        var err = dx - dy;
+
+        while (true) {
+            setPixel(imageData, x0, y0, color); // Do what you need to for this
```
Comment (Bios-Marcel) [inline lobby.html]: What does this comment mean?

File: lobby.html [hunk 6]
```diff
@@ -533,7 +559,154 @@
...
+                    circleData[x][y] = 2;
+                } else {
+                    circleData[x][y] = 1;
+                }
+            }
+        }
+
+        return circleData;
+    }
+
```
Comment (Bios-Marcel) [inline lobby.html]: Whitespace :o

**Comment by Bios-Marcel** [comment]
Another issue is that the fill bucket not only reacts to "Mouse released" and you can accidentally fill things you don't want to fill. I will solve this issue myself tho after the PR is merged, so don't worry, I am just documenting it here.

I've also seen that the fill uses the bresenheim algorithm. That seems unnecessary, as those are two different things in my eyes. Maybe this could be a potential performance improvement, as there is a noticable lag when filling. This might be much worse on weak hardware :man_shrugging:

**Comment by Bios-Marcel** [comment]
Another follow up issue will be making the fill tool optional, since that way people can fallback to having nice smooth lines :D   Like damn ... not having antialiasing takes some time to get used to!
I'll just be reusing the old code for that ;)

File: lobby.html (hunk 1 above)
Comment (bttrs) [reply lobby.html]: so this is a really ugly leftover that I quickly implemented to make it work and then forgot about.
Will deuglify this.

File: lobby.html (hunk 6 above)
Comment (bttrs) [reply lobby.html]: you meant that few lines after that closing parantheses right? :D

**Comment by bttrs** [comment]
sorry about that branch.

Yes the lag is sometimes quite noticable. on all devices i tested it on it was fine enough though (for me at least).

**Comment by bttrs** [comment]
may i ask you whats the difference between the old code and this? I am not noticing any big differences in behaviour apart from the lines looking raw due to missing anti aliasing.

**Comment by Bios-Marcel** [comment]
The performance problems seem to be related to garbage collection. I'll try fixing this myself, it's okay.

Also, I am not sure what you mean with the old code question.

**Comment by Bios-Marcel** [comment]
I actually just found a small library. It seems pretty simple. https://github.com/binarymax/floodfill.js

The cool thing is, that it has a tolerance for anti-aliasing and is super fast.

**Comment by Bios-Marcel** [comment]
I might actually try using that library and bringing back the anti-aliasing. And see how it compares in both performance and visuals.

**Comment by Bios-Marcel** [comment]
Well fuck, the anti-aliasing tolerance is broken, but it's faster either way :D

So, I am gonna use the UI you made, the bresenheim algorithm you wrote and floodfill.js for filling :+1:

**Comment by bttrs** [comment]
dammit, would have been awesome to remain using anti-aliasing :D

**Comment by Bios-Marcel** [comment]
Yes, it deffo looks better, but well :man_shrugging: 

I had another idea, which is having antialiasing in the frontend and having the real drawing in the backend, on which we then operate. But that'd be a bit crazy and slow i guess

**Comment by Bios-Marcel** [comment]
lmfao, the script i found doesn't work with the color black

**Pull request merged**